Implement getAttributeType and getPropertyType functions on TrustedTypePolicyFactory

lukewarlow opened the pull request:
#### 6b79732d9cb95141673959f934689029a8cf6801
<pre>
Implement getAttributeType and getPropertyType functions on TrustedTypePolicyFactory
<a href="https://bugs.webkit.org/show_bug.cgi?id=270080">https://bugs.webkit.org/show_bug.cgi?id=270080</a>

Reviewed by Ryosuke Niwa and Darin Adler.

Implements getAttributeType and getPropertyType metadata functions.
Also adds missing test coverage for SVGScript href attribute handling.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-namespace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-svg-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getPropertyType-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
(WebCore::TrustedTypePolicyFactory::getPropertyType const):
* Source/WebCore/dom/TrustedTypePolicyFactory.idl:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/276310@main">https://commits.webkit.org/276310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25ea46a23ddd6b3b8f95683f5344691acc905d79

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46419 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46630 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40051 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46285 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36277 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37886 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17295 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17577 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38975 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2040 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37352 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40159 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48236 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43576 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18985 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15558 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43137 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20369 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41857 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9849 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20578 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50623 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19996 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10212 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata-expected.txt:
```diff
@@ -16,22 +16,22 @@ PASS Test assignment of TrustedScript on madeup.id
-PASS Test assignment of string on madeup.setAttribute(onerror,..)
-PASS Test assignment of TrustedHTML on madeup.onerror
-PASS Test assignment of TrustedHTML on madeup.setAttribute(onerror,..)
+FAIL Test assignment of string on madeup.onerror assert_throws_js: throws function "_ => { element[property] = value; }" did not throw
```
These failures are caused by the fact that the helper method is now implemented but the assignment isn't yet protected.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    AtomString sink;
+};
+
+
```
Nit: Two blank lines here.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
-String TrustedTypePolicyFactory::getAttributeType(const String& tagName, const String& attribute, const String& elementNamespace, const String& attrNamespace) const
+struct AttributeTypeListItem {
+    AtomString element;
+    AtomString elementNs;
```
elementNS or spell out ElementNamespace.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    AtomString element;
+    AtomString elementNs;
+    AtomString attribute;
+    AtomString attributeNs;
```
Ditto.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+
+static const Vector<AttributeTypeListItem> attributeTypeVector()
+{
+    static MainThreadNeverDestroyed<Vector<AttributeTypeListItem>> result;
```
Can't we juts do: `result = {...}` instead?

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    Vector<PropertyTypeListItem> table = propertyTypeVector();
+
+    for (const auto& entry : table) {
+        bool elementMatches = entry.element == AtomString(localName) && entry.elementNs == elementNs;
```
This is awfully inefficient. We should be using ElementName instead.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    for (const auto& entry : table) {
+        bool elementMatches = entry.element == AtomString(localName) && entry.elementNs == elementNs;
+        bool entryMatches = (elementMatches
+            || entry.element == AtomString("*"_s))
```
Let's not convert `"*"` to `AtomString` in every iteration. We should just use null string instead.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    if (property.startsWith("on"_s))
+        return "TrustedScript"_s;
+
+    Vector<PropertyTypeListItem> table = propertyTypeVector();
```
What's the point of creating this Vector? Why can't just hard-code the list of elements & attributes using switch statements?

lukewarlow replied:
Not sure how to do it with switches but changed it to use if statements instead.

lukewarlow replied:
@rniwa this should be ready for a re-review.

rniwa reviewed (changes requested):
I don't think we should land this as is.

darinadler commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,21 +101,59 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
 }
 
-String TrustedTypePolicyFactory::getAttributeType(const String& tagName, const String& attribute, const String& elementNamespace, const String& attrNamespace) const
+String TrustedTypePolicyFactory::getAttributeType(const String& tagName, const String& attributeParam, const String& elementNamespace, const String& attrNamespace) const
```
In WebKit coding style we normally avoid abbreviations like "param" and instead use words like "parameter".

darinadler commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,21 +101,59 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+    AtomString attrNS = attrNamespace.isEmpty() ? nullAtom() : AtomString(attrNamespace);
+
+    const QualifiedName element(nullAtom(), AtomString(localName), elementNS);
+    const QualifiedName attribute(nullAtom(), AtomString(attributeName), attrNS);
```
Not sure the `const` adds anything here. There are lots of local variables that we don’t change. I would leave it out.

darinadler commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -35,9 +35,9 @@ interface TrustedTypePolicyFactory {
     readonly attribute TrustedScript emptyScript;
 
-    DOMString? getAttributeType(DOMString tagName, DOMString attribute, optional DOMString elementNamespace = "", optional DOMString attrNamespace = "");
+    DOMString? getAttributeType(DOMString tagName, DOMString attribute, optional [LegacyNullToEmptyString] DOMString elementNamespace = "", optional [LegacyNullToEmptyString] DOMString attrNamespace = "");
```
Very surprising that a new function requires this legacy attribute. Do you know why?

lukewarlow replied:
I was confused why it was needed aswell but turns out it isn't the DOMString just needs a `?` instead.

darinadler commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -35,9 +35,9 @@ interface TrustedTypePolicyFactory {
+    DOMString? getAttributeType(DOMString tagName, DOMString attribute, optional [LegacyNullToEmptyString] DOMString elementNamespace = "", optional [LegacyNullToEmptyString] DOMString attrNamespace = "");
 
-    DOMString? getPropertyType(DOMString tagName, DOMString property, optional DOMString elementNamespace = "");
+    DOMString? getPropertyType(DOMString tagName, DOMString property, optional [LegacyNullToEmptyString] DOMString elementNamespace = "");
```
Very surprising that a new function requires this legacy attribute. Do you know why?

lukewarlow replied:
ditto

darinadler reviewed (approved):
Looks OK. Not sure it’s absolutely correct because of the peculiar handling of null.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -100,21 +101,59 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
 }
 
-String TrustedTypePolicyFactory::getPropertyType(const String& tagName, const String& attribute, const String& elementNamespace) const
+String TrustedTypePolicyFactory::getPropertyType(const String& tagName, const String& property, const String& elementNamespace) const
```
Looks like a spec has a typo?
https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getpropertytype
should probably say "If interface has an IDL [attribute](https://webidl.spec.whatwg.org/#dfn-attribute) member which identifier is *property*," not "*attribute*",

lukewarlow replied:
Good spot, I've got a spec PR that rewrites this function I'll check that it's correct on my PR.

lukewarlow committed:
Implement getAttributeType and getPropertyType functions on TrustedTypePolicyFactory
https://bugs.webkit.org/show_bug.cgi?id=270080

Reviewed by Ryosuke Niwa and Darin Adler.

Implements getAttributeType and getPropertyType metadata functions.
Also adds missing test coverage for SVGScript href attribute handling.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-namespace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-svg-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getPropertyType-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
(WebCore::TrustedTypePolicyFactory::getPropertyType const):
* Source/WebCore/dom/TrustedTypePolicyFactory.idl:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt: Added.

Canonical link: https://commits.webkit.org/276310@main
